src/main/java/frc/robot/subsystems/vision/VisionIO.java

```diff
@@ -0,0 +1,37 @@
+// Copyright (c) FIRST and other WPILib contributors.
+// Open Source Software; you can modify and/or share it under the terms of
+// the WPILib BSD license file in the root directory of this project.
+
+package frc.robot.subsystems.vision;
+
+import edu.wpi.first.math.geometry.Pose3d;
+import frc.robot.subsystems.vision.Vision.VisionConstants;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import org.photonvision.EstimatedRobotPose;
+import org.photonvision.targeting.PhotonTrackedTarget;
+
+/** Add your docs here. */
+public interface VisionIO {
+
+  public static class VisionIOInputs {
+    public double timestamp = 0.0;
+    public double latency = 0.0;
+    public List<PhotonTrackedTarget> targets =
+        new ArrayList<>(); // TODO make protobuf work whenever that happens
+    public double numTags = 0; // TODO why isn't this just targets.size()?
+    public Pose3d coprocPNPPose = new Pose3d();
+    public Pose3d[] targetPose3ds;
+    public VisionConstants constants;
+  }
+
+  public default void updateInputs(VisionIOInputs inputs) {}
```

Comment:
Unless we've got a good reason to do it we should avoid default implementations doing nothing. It makes it easy to forget to override and then you've got a method doing nothing when you didn't intend that.

The `getName` version is pretty innocuous, but I'd cut it from all the other instances.

Reply:
what should i do instead? i'm not sure how else to call it from the io in Vision.java

Reply:
I'm not sure what you're asking. `VisionIO` is an interface. If you reference it somewhere, you're saying that the object in question implements that interface and you can call that set of methods on it, but doesn't say what the class of the object is. `Vision` takes a `VisionIO` and when you construct it in `SwerveSubsystem.createRealCameras` you make a concrete class that implements `VisionIO` (`VisionIOReal`) so you're good. The default null implementation just makes it so if you make a `class Foo implements VisionIO` and you don't explicitly implement that method, it comes with a default implementation. That makes sense sometimes, but a null implementation is a foot-gun because it's easy to accidentally end up with a method that does nothing by default and the compiler doesn't force you to notice.

Reply:
So long story short, cut the `default` and the `{}` and add a `;` at the end.

Reply:
fixed!

Reply:
You can click the "resolved" button and the comment chain will get archived.